### PR TITLE
Allow `qalloc_p` to lower to a dynamic number of wires

### PR DESCRIPTION
**Pull request by paul0403**
**Context:**
Although the quantum.alloc operation in the mlir layer [allows both a static and dynamic number of wires](https://github.com/PennyLaneAI/catalyst/blob/main/mlir/include/Quantum/IR/QuantumOps.td#L117), in the frontend when lowering `qalloc_p`, the static option is always used. This means we cannot lower a circuit with a dynamic number of wires to mlir. 

**Description of the Change:**
Update `qalloc_p` lowering so it can handle a dynamic number of wires.
Correspondingly, when tracing, the dynamic jaxpr tracer is passed into the alloc primitive when appropriate.

**Benefits:**
Catalyst can compile circuits with a dynamic number of wires.

**Possible Drawbacks:**
Measurements whose return shape depend on number of wires are still static. This is because this PR is step 1 of splitting #1532 .

**Related GitHub Issues:**

[sc-84672]

### CODE REVIEW

File: frontend/catalyst/jax_primitives.py [hunk 1]
```diff
@@ -801,21 +801,21 @@ def _qalloc_abstract_eval(size):
     return AbstractQreg()
 
 
-def _qalloc_lowering(jax_ctx: mlir.LoweringRuleContext, size_value: ir.Value):
+def _qalloc_lowering(jax_ctx: mlir.LoweringRuleContext, size_value: int | ir.Value):
```
Comment (dime10) [inline frontend/catalyst/jax_primitives.py:804]: When is this an int?

Comment (paul0403) [reply frontend/catalyst/jax_primitives.py:804]: Good catch! I forgot to remove it :sweat_smile:

File: frontend/catalyst/jax_primitives.py [hunk 2]
```diff
@@ -805,16 +805,17 @@ def _qalloc_lowering(jax_ctx: mlir.LoweringRuleContext, size_value: ir.Value):
...
-    assert ir.DenseIntElementsAttr.isinstance(size_value_attr)
-    size = ir.DenseIntElementsAttr(size_value_attr)[0]
-
     qreg_type = ir.OpaqueType.get("quantum", "reg", ctx)
-    i64_type = ir.IntegerType.get_signless(64, ctx)
-    size_attr = ir.IntegerAttr.get(i64_type, size)
+    if isinstance(size_value.owner, ir.Operation) and size_value.owner.name == "stablehlo.constant":
+        size_value_attr = size_value.owner.attributes["value"]
+        assert ir.DenseIntElementsAttr.isinstance(size_value_attr)
+        size = ir.DenseIntElementsAttr(size_value_attr)[0]
```
Comment (mehrdad2m) [inline frontend/catalyst/jax_primitives.py:812]: I am curious to know what happens if a negative wire is given. Does frontend check if the wire number is positive or is that handled by runtime?

File: frontend/catalyst/device/qjit_device.py [hunk 3]
```diff
@@ -309,7 +310,11 @@ def __init__(self, original_device):
         for key, value in original_device.__dict__.items():
             self.__setattr__(key, value)
 
-        check_device_wires(original_device.wires)
+        if original_device.wires is None:
```
Comment (mehrdad2m) [inline frontend/catalyst/device/qjit_device.py:313]: Is it possible still to do this check inside `check_device_wires` and maybe check for `is_dynamic_wires` after this check inside the function and return if in dynamic wire mode?

File: frontend/catalyst/jax_extras/tracing.py [hunk 4]
```diff
@@ -999,3 +999,16 @@ def bind_flexible_primitive(primitive, flexible_args: dict[str, Any], *dyn_args,
             dyn_args += (flex_arg_value,)
 
     return primitive.bind(*dyn_args, **static_args)
+
+
+def is_dynamic_wires(wires: pennylane.wires.Wires):  # pylint: disable=undefined-variable
+    """
+    Checks if a pennylane Wires object corresponds to a concrete number
```
Comment (mehrdad2m) [inline frontend/catalyst/jax_extras/tracing.py:1006]:
```suggestion
    Checks if a pennylane :class:`qml.wires.Wires <pennylane.wires.Wires>` object corresponds to a static number
```
Not sure if this correct but I guess there is a way to link to the [wires](https://docs.pennylane.ai/en/stable/code/api/pennylane.wires.Wires.html) page in docs

File: frontend/catalyst/jax_extras/tracing.py [hunk 5]
```diff
@@ -999,3 +999,16 @@ def bind_flexible_primitive(primitive, flexible_args: dict[str, Any], *dyn_args,
...
+    of wires or a dynamic number of wires.
+
+    If the number of wires is concrete, the Wires object contains a list of wire labels,
```
Comment (mehrdad2m) [inline frontend/catalyst/jax_extras/tracing.py:1009]:
```suggestion
    If the number of wires is static, the Wires object contains a list of wire labels,
```
static is maybe less vague in this context.

File: frontend/catalyst/jax_extras/tracing.py [hunk 6]
```diff
@@ -999,3 +999,16 @@ def bind_flexible_primitive(primitive, flexible_args: dict[str, Any], *dyn_args,
             dyn_args += (flex_arg_value,)
 
     return primitive.bind(*dyn_args, **static_args)
+
+
+def is_dynamic_wires(wires: pennylane.wires.Wires):  # pylint: disable=undefined-variable
```
Comment (mehrdad2m) [inline frontend/catalyst/jax_extras/tracing.py:1004]: From the docstring explanation it can be interpreted the the result of this function is designed to be True -> dynamic, False -> static. However, this may not be the case if we have a list of multiple tracers. Maybe here this check should be. a bit could be a bit more aggressive and check the size in check_device_wires. The other options is to change the docstring. What do you think?

File: frontend/test/pytest/test_measurement_dynamic_shapes.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+def test_dynamic_wires_scalar_readouts(readout, backend, capfd):
+    """
+    Test that a circuit with dynamic number of wires can be executed correctly.
+
+    As a unit test for allocating a dynamic number of wires, we use measurements
+    whose shape do not depend on the number of wires, i.e. expval and var
+    """
+
+    def ref(num_qubits):
+        print("compiling...")
```
Comment (mehrdad2m) [inline frontend/test/pytest/test_measurement_dynamic_shapes.py:118]:
```suggestion
```

**Review by mehrdad2m — COMMENTED** [review]
Looks great @paul0403 🚀

File: frontend/catalyst/jax_primitives.py (hunk 2 above)
Comment (paul0403) [reply frontend/catalyst/jax_primitives.py:812]: Unmeaningful wire numbers are handled by runtime: https://github.com/PennyLaneAI/catalyst/blob/main/runtime/lib/capi/RuntimeCAPI.cpp#L1013

File: frontend/test/pytest/test_measurement_dynamic_shapes.py (hunk 7 above)
Comment (paul0403) [reply frontend/test/pytest/test_measurement_dynamic_shapes.py:118]: I need this print to check that recompilation is not happening between the two runs. See the `out.count("compiling...")` assertion in the test below.

File: frontend/test/lit/test_measurements.py [hunk 8]
```diff
@@ -471,6 +471,38 @@ def expval10(x: float, y: float):
...
+def expval11(num_qubits):
+    # CHECK: func.func public @circ(%arg0: tensor<i64>) -> tensor<f64>
+
+    # CHECK-DAG: [[one:%.+]] = stablehlo.constant dense<1> : tensor<i64>
+    # CHECK-DAG: [[two:%.+]] = stablehlo.constant dense<2> : tensor<i64>
+    # CHECK-DAG: [[nSub1:%.+]] = stablehlo.subtract %arg0, [[one]] : tensor<i64>
+    # CHECK-DAG: [[nSub2:%.+]] = stablehlo.subtract %arg0, [[two]] : tensor<i64>
+
+    # CHECK: [[deten_nQubits:%.+]] = tensor.extract %arg0[] : tensor<i64>
+    # CHECK: [[reg:%.+]] = quantum.alloc([[deten_nQubits]]) : !quantum.reg
```
Comment (dime10) [inline frontend/test/lit/test_measurements.py:485]: I was confused by the ordering, shouldn't these check statements (except the first one) be inside the circ function (for clarity)?

File: frontend/test/pytest/test_measurement_dynamic_shapes.py (hunk 7 above)
Comment (paul0403) [reply frontend/test/pytest/test_measurement_dynamic_shapes.py:118]: I don't know if we have any fixture/utility to check this.

File: frontend/test/pytest/test_measurement_primitives.py
Comment (dime10) [inline frontend/test/pytest/test_measurement_primitives.py:1]: Why are you deleting these?

**Review by dime10 — APPROVED** [review]
Straight forward PR, thank you 👍

File: frontend/catalyst/jax_primitives.py (hunk 2 above)
Comment (mehrdad2m) [reply frontend/catalyst/jax_primitives.py:812]: sounds good. It is not very important in this context but I think it is useful in general to check these kinds of trivial checks both in frontend and runtime to avoid any risk of future mistransformation of an unmeaningful value to a meaningful value by mistake at which point the runtime check would fail to catch the issue.

File: frontend/catalyst/jax_extras/tracing.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (paul0403) [reply frontend/catalyst/jax_extras/tracing.py:1004]: Yes, I did try that and there are circular import issues. I agree this is not the best place to place this function. Let me see what I can do here.

File: frontend/test/pytest/test_measurement_dynamic_shapes.py (hunk 7 above)
Comment (mehrdad2m) [reply frontend/test/pytest/test_measurement_dynamic_shapes.py:118]: Ah I see. Thanks for clarifying

File: frontend/catalyst/jax_primitives.py (hunk 2 above)
Comment (paul0403) [reply frontend/catalyst/jax_primitives.py:812]: Good point! I added it https://github.com/PennyLaneAI/catalyst/pull/1549/commits/916f84b307ed0cb31f772933e2201e7bd9d7dfa3

File: frontend/test/lit/test_measurements.py (hunk 8 above)
Comment (dime10) [reply frontend/test/lit/test_measurements.py:485]: True, I just meant visually 😄

File: frontend/catalyst/jax_extras/tracing.py (hunk 4 above)
Comment (paul0403) [reply frontend/catalyst/jax_extras/tracing.py:1006]: I think we don't need to link this because this file is strictly internal, and the functions are not exposed into documentation.

File: frontend/test/lit/test_measurements.py (hunk 8 above)
Comment (paul0403) [reply frontend/test/lit/test_measurements.py]: Ahh, I see! https://github.com/PennyLaneAI/catalyst/pull/1549/commits/dc3b6b0c1647d9482c9c9f27fb38a6b52a630108

File: frontend/catalyst/device/qjit_device.py [hunk 9]
```diff
@@ -309,7 +310,11 @@ def __init__(self, original_device):
...
+            raise AttributeError("Catalyst does not support device instances without set wires.")
+
+        if not is_dynamic_wires(original_device.wires):
```
[the remark this answers is not in the document]
Comment (paul0403) [reply frontend/catalyst/device/qjit_device.py]: https://github.com/PennyLaneAI/catalyst/pull/1549/commits/d74ace315294c0fa3bae9b89e99183c79cdbd618

File: frontend/catalyst/device/qjit_device.py [hunk 10]
```diff
@@ -309,7 +310,11 @@ def __init__(self, original_device):
...
+            check_device_wires(original_device.wires)
```
[the remark this answers is not in the document]
Comment (paul0403) [reply frontend/catalyst/device/qjit_device.py]: https://github.com/PennyLaneAI/catalyst/pull/1549/commits/d74ace315294c0fa3bae9b89e99183c79cdbd618

File: frontend/catalyst/device/qjit_device.py (hunk 3 above)
Comment (paul0403) [reply frontend/catalyst/device/qjit_device.py]: https://github.com/PennyLaneAI/catalyst/pull/1549/commits/d74ace315294c0fa3bae9b89e99183c79cdbd618

File: frontend/catalyst/jax_extras/tracing.py (hunk 6 above)
Comment (paul0403) [reply frontend/catalyst/jax_extras/tracing.py:1004]: Good point, I added a check that if the wires object has more than one entry then they must not be tracers https://github.com/PennyLaneAI/catalyst/pull/1549/commits/084398750354e60164ab9e75f805362338e4e9b2

[the remark this answers is not in the document]
Comment (paul0403) [reply frontend/catalyst/jax_extras/tracing.py:1004]: So importing anything in `catalyst.device` in jax_tracers will lead to circular imports. On the other hand, we can do a lengthy `catalyst.device.qjit_device.is_dynamic_wires`  in `jax_tracers`. 

This is a trade-off between slightly longer code and a function being in a non-appropriate place. I prefer the longer call, to be honest.

[the remark this answers is not in the document]
Comment (paul0403) [reply frontend/catalyst/jax_extras/tracing.py]: https://github.com/PennyLaneAI/catalyst/pull/1549/commits/12f890b9c4077f7e3396a1a23ba264a1933df4ab

File: frontend/catalyst/device/qjit_device.py [hunk 11]
```diff
@@ -532,16 +535,33 @@ def get_device_capabilities(device) -> DeviceCapabilities:
...
+        if not all(isinstance(wire, int) for wire in wires.labels):
+            raise AttributeError("Catalyst requires continuous integer wire labels starting at 0.")
 
-    if not wires.labels == tuple(range(len(wires))):
-        raise AttributeError("Catalyst requires continuous integer wire labels starting at 0.")
+        if not wires.labels == tuple(range(len(wires))):
+            raise AttributeError("Catalyst requires continuous integer wire labels starting at 0.")
+    else:
+        assert wires[0].shape == ()
+        assert wires[0].dtype == "int64"
```
Comment (dime10) [inline frontend/catalyst/device/qjit_device.py:567]: Given that this is checking a user provided value, I think these should be errors telling the user what they did wrong :) 

We could also consider being a little forgiving with the shape, since the scalar extract function already supports both `()` and `(1,)` :)

**Review by mehrdad2m — APPROVED** [review]

File: frontend/catalyst/device/qjit_device.py (hunk 11 above)
Comment (dime10) [reply frontend/catalyst/device/qjit_device.py]:
```py
        assert len(wires) == 1
        assert wires[0].shape in ((), (1,))
```
The second assert also needs to be an error, since the only thing we've checked so far is that the length of wires is 1 and that the first element is a tracer.

Comment (paul0403) [reply frontend/catalyst/device/qjit_device.py]: So I actually tried this, by sending in
```python
@pytest.mark.parametrize("wires", [(1, 2)])
def test_wrong_wires_argument(backend, wires):
```
in the pytest. However, this got sent to the previous check, i.e. the `Catalyst requires continuous integer wire labels starting at 0.`. This is because the pytree is falttened in the wires object (unsure), and before reaching the check it becomes 
```
Wires([Traced<ShapedArray(int64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>, Traced<ShapedArray(int64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>])
```
i.e. a length-2 object, instead of a length-1 object with the object having a shape of (2,). Any thoughts on how to deal with this?

Comment (dime10) [reply frontend/catalyst/device/qjit_device.py]: I think this should trigger if the user specified `wires=jnp.array([0,1])`. If not we're probably fine with the assertion :)

Comment (dime10) [reply frontend/catalyst/device/qjit_device.py]: Just tested it, I guess the device will split the array up somehow, even if they are tracers. All good then 👍
